exploration/semantic-skeletons.md

```diff
@@ -0,0 +1,266 @@
+- [ICU4X Field Set Builder \(more JS-like\)](https://unicode-org.github.io/icu4x/rustdoc/icu/datetime/fieldsets/builder/struct.FieldSetBuilder.html)
+
+
+Semantic skeletons are not the first attempt to provide this functionality.
+Previous skeleton mechanisms ("classical skeletons") used 
+collections of field options (as in `Intl.DateTimeFormat`)
+or a microsyntax (as in ICU4J).
+
+The `Intl.DateTimeFormat` skeletons consist of "option bags"
+such as `{ year: "numeric", month: "short", day: "numeric" }`
+in which the user specifies the field and its width.
+Only fields appearing in the options appear in the formatted date/time value.
+
+The ICU MessageFormat "classical skeleton" microsyntax uses strings supplied by the developers.
+These strings specify the fields and field lengths that should appear in the formatted value.
+See [here](https://unicode-org.github.io/icu/userguide/format_parse/datetime/#date-field-symbol-table)
+The system then uses the string to perform date/time pattern generation,
+arranging the specified fields in the correct order,
+selecting locale-appropriate separators,
+and producing a "picture string" that can be consumed by date/time formatters
+such as `java.text.SimpleDateFormat`.
+
+## Use-Cases
+
+_What use-cases do we see? Ideally, quote concrete examples._
+
+As a developer, I want to format date, time, or date/time values to show specific fields
+with specific appearance without having to learn a complex microsyntax
+or modify my code or formatting directions for each locale.
+
+As a translator, I want to understand what output a given date/time placeholder will produce 
+in my language.
+
+As a translator, I don't want to have to "translate" or modify a date/time placeholder to suit
+my language's needs.
+I should trust that the placeholder will produce appropriate results for my language.
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+Users want the most intuitive formats and outcomes to be the defaults.
+They should not have to coerce or convert normal date/time types in order to do simple operations.
+
+1. It should be possible to format operands consisting of locally-relevant date/time types, including:
+   - Temporal values such as `java.time` or JS `Temporal` values,
+   - incremental time types ("timestamps")
+     (e.g. milliseconds since epoch times such as `java.util.Date`, `time_t`, JS `Date`, etc.),
+   - field-based time types
+     (e.g. those that contain separate values per field type in a date/time, such as a year-month),
+   - [floating time](https://www.w3.org/TR/timezone/#dfn-floating-time) values
+     (e.g. those that are not tied to a specific time zone, variously called local/plain/civil times),
+   -  or other local exotica (Java `Calendar`, C `tm` struct, etc.)
+1. Date/time formatters should not permit users to format fields that don't exist in the value
+   (e.g. the "month" of a time, the "hour" of a date)
```

Comment:
This presumably should not be understood to mean that e.g. epoch milliseconds ought not be formattable as year-month-date, given that the value only contains milliseconds, yes?

```suggestion
1. Date/time formatter options should be able to impose restrictions on acceptable values,
   such as formatting a month requiring a date, and formatting an hour requiring a time.
```

Reply:
They're not really "formatter options" though, are they? `{$timeOnlyValue :date}` == Bad Operand?

```suggestion
1. Date/time formatting functions and their options should be able to impose restrictions on acceptable values,
   such as formatting a month requiring a value that includes the month 
   or formatting an hour requiring requiring a time.
1. Date/time formatting functions and their options should be able to emit errors 
   when trying to format fields that cannot be computed from the operand value
   such as the "month" of a time-only value (`{|03:45Z| :date format=YMD}`)
   or the "hour" of a date-only value (`{|2025-04-23| :time format=HM}`).
```

Reply:
Your suggested second requirement smells like scope creep to me. I understand this design doc to be about function options, not about enabling or restricting a function handler from emitting errors.

We don't actually impose any restriction on when a function handler can emit an error (just on what errors may be emitted when), we already allow an implementation with an options-basket or even classical-skeleton option style to reject option combinations that could not be expressed as a semantic skeleton.

This design doc should be focused on the ergonomics and readability of the different options styles, and not on implementation details.